kpatch-build/kpatch-elf.c

```diff
@@ -726,6 +753,119 @@ void kpatch_reindex_elements(struct kpatch_elf *kelf)
 	}
 }
 
+void kpatch_mark_grouped_sections(struct kpatch_elf *kelf)
+{
+	struct section *groupsec, *sec;
+	struct group_section *gsec;
+	struct symbol *tsym;
+	unsigned int *data, *end, i;
+
+	INIT_LIST_HEAD(&kelf->groupsec);
+
+	list_for_each_entry(groupsec, &kelf->sections, list) {
+		i = 0;
+		if (groupsec->sh.sh_type != SHT_GROUP)
+			continue;
+		/*
+		 * a. Maintain the section index in struct section and struct
+		 *    group_section. This is later used to compare if the reindexed
+		 *    section and group_section matches.
+		 * b. Maintain the secnames, symnames to later update
+		 *    group section members sh_link and sh_info.
+		 */
+		groupsec->groupindex = groupsec->index;
+		data = groupsec->data->d_buf;
+		end = groupsec->data->d_buf + groupsec->data->d_size;
+		data++; /* skip first flag word (e.g. GRP_COMDAT) */
+		gsec = malloc(sizeof(*gsec));
+		if (!gsec)
+			ERROR("malloc");
+		gsec->groupindex = groupsec->index;
+		/* Maintain names of all section link. Use this info after reindexing */
+		gsec->secnames = malloc((groupsec->data->d_size/sizeof(*data)) * sizeof(char **));
+		if (!gsec->secnames)
+			ERROR("malloc");
+		gsec->symnames = malloc((groupsec->data->d_size/sizeof(*data)) * sizeof(char **));
+		if (!gsec->symnames)
+			ERROR("malloc");
+
+		while (data < end) {
+			sec = find_section_by_index(&kelf->sections, *data);
+			if (!sec)
+				ERROR("group section not found");
+			sec->grouped = 1;
+			log_debug("marking section %s (%d) as grouped\n",
+				  sec->name, sec->index);
+
+			gsec->secnames[i] = strdup(sec->name);
+			tsym = find_symbol_by_index(&kelf->symbols, groupsec->sh.sh_info);
+			if (!tsym)
+				ERROR("symbol not found");
+			gsec->symnames[i] = strdup(tsym->name);
+			data++;
+			i++;
+		}
+		list_add_tail(&gsec->list, &kelf->groupsec);
+	}
+}
+
+/*
+ * kpatch-build reindexes the sections and symbols. When these are reindexed,
+ * group section needs reindexing as well.
+ *
+ * In s390x, expolines like __s390_indirect_jump_r1 are the referenced in
+ * .group sections.
+ */
+void kpatch_reindex_group_sections(struct kpatch_elf *kelf)
+{
+	struct section *sec, *subsec, *tsec;
+	struct symbol *tsym;
+	struct group_section *groupsec;
+	unsigned int *data, *end, i = 0;
+	unsigned int *newdata;
+
+	list_for_each_entry(groupsec, &kelf->groupsec, list) {
+		i = 0;
+		list_for_each_entry(sec, &kelf->sections, list) {
+			/* section's groupindex is set, only if it is a group section */
+			if (sec->sh.sh_type != SHT_GROUP)
+				continue;
+			if (sec->groupindex != groupsec->groupindex)
+				continue;
+			data = sec->data->d_buf;
+			end = sec->data->d_buf + sec->data->d_size;
+			newdata = malloc(sec->data->d_size);
```

Comment:
Likewise for these allocations.

Reply:
Cleanup of this allocation performed in kpatch_free_groupsec.